certora/specs/CER-11-Controllers/CER-12-Controllers-number.spec

```diff
@@ -7,20 +7,63 @@ Each Account can have at most one Controller Vault enabled at a time unless it's
 
 methods {
     function numOfController(address account) external returns(uint8) envfree;
+    function getExecutionContextAreChecksDeferred() external returns(bool) envfree;
+    function containsStatusCheckFor(address account) external returns(bool) envfree;
 }
 
 /**
  * Check that we never have more than one controller.
  *
- * It does not work for batch(), as it requires rather complex reasoning. Some
+ * It does not work for batch() or call(), as it requires rather complex reasoning. Some
  * pieces necessary for this reasoning:
  * - the account status checks verifies that there is only a single controller.
  * - the only place that adds a controller issues an account status check for
  *   the respective account, either immediately or by registering it to the set.
- * - at the end of batch(), the execution context is restored.
+ * - whenever the setDeferredChecks bit in `executionContext` is cleared,
+ *   we make sure that we run the account status checks for
+ *   all accounts registered for a status check.
+ * 
+ * We prove this separately for enableController as we need to break it into
+ * cases (when checks are deferred vs not). This is done in the 3 rules that
+ * follow this one.
  */
 invariant onlyOneController(address a)
     numOfController(a) <= 1
     filtered { // TODO: make this work for batch as well
         f -> f.selector != sig:batch(IEVC.BatchItem[] calldata).selector
+            && f.selector != sig:call(address, address, uint256, bytes calldata).selector
+            && f.selector != sig:enableController(address, address).selector
     }
+
+// For enableController, we can only check this invariant directly if 
+// account status checks are not deferred. To cover the deferred case,
+// we check that when account checks are deferred, the enableController
+// call will enqueue a check for the relevant account using the rule
+// `enableControllerEnqueuesStatusCheckWhenDeferred`
+invariant onlyOneControllerEnableController(address a)
+    numOfController(a) <= 1 && !getExecutionContextAreChecksDeferred()
+    filtered {
+        f -> f.selector == sig:enableController(address, address).selector
+    }
+
+
+// This ensures enableController will enqueue a status
+// check when status checks are deferred
+rule enableControllerEnequeusStatusCheckWhenDeferred {
+    env e;
+    address account;
+    address vault;
+    require getExecutionContextAreChecksDeferred();
+    enableController(e, account, vault);
+    assert containsStatusCheckFor(account) == true;
+}
+
+// This directly specifies that checkAccountStatusInternal
+// will force numOfController for the checked account to be
+// less or equal to one to cover deferred checks.
+rule checkAccountStatusForcesNumControllersLEOne {
+    env e;
+    address account;
+    checkAccountStatus(e, account);
+    assert numOfController(account) <= 1;
+}
```

Comment:
should we have a rule that make sure that if getExecutionContextAreChecksDeferred is turned on then checkAccountStatus must be called? 
Under which condition does getExecutionContextAreChecksDeferred is turned on, do we have a rule for that?

Reply:
it will be verified as per these requirements:
https://linear.app/euler-labs/issue/CER-75/account-and-vault-status-checks-execution
https://linear.app/euler-labs/issue/CER-37/account-and-vault-status-checks-deferred
https://linear.app/euler-labs/issue/CER-36/account-and-vault-status-checks-immediate

Reply:
@kasperpawlowski Do you think your contract implementation that ensures that all the deferred vault status checks are eventually executed (all the ones enqueued in vaultStatusChecks will eventually be performed) by some time? I think those 3 rules do not capture this.

It seems like it might -- I see that `checkStatusAll` executes all the checks in the set, that this is called by `restoreExecutionContext`, and `restoreExecutionContext` is called at the end of `batch`, `call`, and `controlCollateral`. Are those 3 calls the only place where `deferStatusChecks` can be turned on and off? If so I think it may be valuable for security to try to capture this with another rule. (Though I do suspect it will be a tricky one for us to prove).

What do you think?

Reply:
I added 2 more issues (set to low priority for now) so that we remember to look into this in case we can find the time:

https://linear.app/euler-labs/issue/CER-106/account-and-vault-status-checks-deferred-liveness
https://linear.app/euler-labs/issue/CER-107/account-and-vault-status-checks-deferred-scoping

Reply:
@andrew-certora I updated CER-73 to capture what you wrote in CER-107. I think it better fits there. if so, CER-107 could be removed
https://linear.app/euler-labs/issue/CER-73/checks-deferrable-calls-execution

as for the liveness check (CER-106), I agree, it will be helpful to verify. thank you!